modules/caddytls/certgetters.go

```diff
@@ -0,0 +1,169 @@
+package caddytls
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/certmagic"
+	"tailscale.com/client/tailscale"
+)
+
+func init() {
+	caddy.RegisterModule(Tailscale{})
+	caddy.RegisterModule(HTTPCertGetter{})
+}
+
+// Tailscale is a module that can get certificates from the local Tailscale process.
+type Tailscale struct {
+	// Whether to cache returned certificates in Caddy's in-memory certificate cache.
+	// If true, Tailscale will only be asked for a certificate if it does not already
+	// exist in Caddy's cache, or if it is nearing expiration.
+	Cache bool `json:"cache,omitempty"`
+}
+
+// CaddyModule returns the Caddy module information.
+func (Tailscale) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "tls.get_certificate.tailscale",
+		New: func() caddy.Module { return new(Tailscale) },
+	}
+}
+
+func (ts Tailscale) GetCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, bool, error) {
+	cert, err := tailscale.GetCertificate(hello)
+	return cert, ts.Cache, err
+}
+
+// UnmarshalCaddyfile deserializes Caddyfile tokens into ts.
+//
+//     ... tailscale {
+//         cache
+//     }
+//
+func (ts *Tailscale) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	for d.Next() {
+		if d.NextArg() {
+			return d.ArgErr()
+		}
+		for nesting := d.Nesting(); d.NextBlock(nesting); {
+			switch d.Val() {
+			case "cache":
+				if ts.Cache {
+					return d.Errf("caching is already enabled")
+				}
+				if d.NextArg() {
+					return d.ArgErr()
+				}
+				ts.Cache = true
+			default:
+				return d.Errf("unrecognized tailscale property: %s", d.Val())
+			}
+		}
+	}
+	return nil
+}
+
+// HTTPCertGetter can get a certificate via HTTP(S) request.
+type HTTPCertGetter struct {
+	// The URL from which to download the certificate. Required.
+	//
+	// The URL will be augmented with query string parameters taken
+	// from the TLS handshake:
+	//
+	// - server_name: The SNI value
+	// - signature_schemes: Comma-separated list of hex IDs of signatures
+	// - cipher_suites: Comma-separated list of hex IDs of cipher suites
+	//
+	// To be valid, the response must be HTTP 200 with ... TODO: body format?
```

Comment:
Do we need it to return the full chain? Maybe requiring JSON is best, more flexible ultimately. If someone else wants another format they can make a plugin 🤷‍♂️

Reply:
Yeah, the whole `tls.Certificate` struct, including private key (yikes). I was thinking either JSON or multipart form. Leaning towards JSON though.

Reply:
Could add have an `accepts` option to change what data type Caddy expects to get back, but only if someone needs it/requests support for it.